### DISCUSSION TITLE
Datetime format in some of the reader.

### DISCUSSION THREAD

**Discussion by fernandezc** [spectrochempy/spectrochempy#366] [Developer discussions]
@atravert Arnaud, I am not familiar with the hardware producing the format read by read_quadera. But I notice that the Datetime format is in French!  I guess it is either a choice of the user, or fixed by the regional settings of the computer. If a different format can be obtained by different settings it should be possible to read it: Could you check this in this case and possibly for other software like omnic and opus?

**Comment by fernandezc** [comment]
No in quadera it is not a French format sorry.  However I still wonder if the date format can be changed in these outputs.

**Reply by fernandezc** [reply]
Well, for labspec file, the examples given have date in French. 
I am pretty sure this can be changed on the apparatus... so I started writing a simple parser in case the date is not in the ISO format. 
We could use a third party package for this : dateutil, but there is already too much dependency. I would prefer not to add a new one.

**Reply by fernandezc** [reply]
It's even worse, the format is not the same in different files in ramandata: 

- Activation.txt    -> #Date=	01/05/16 11:27:14
- serie190214-1.txt  -> #Date=	14.02.2019 10:41

Very strange.

**Comment by atravert** [comment]
Here a summary of the date formats currently handled by scp readers : 

Binary formats:
-------------
- Omnic (spa, spg, srs): the acquisition date is given by a timestamp (HFS+32 bit, little endian) from 31/12/1899, 00:00), but i did not invertigate to find out the timezone is (i have to make acquisitions with different timezone settings and find out the changes in the file!). 
- Opus: the date, time and shift with GMT are explicitely indicated with `DAT`, `TIM`, `(GMT+1)` (HEX) keywords 
- grams SPC: well documented in the file format specification, but (i) it does not give seconds and (ii) there is not specification for the time zone - by the way i have to remove (GMT) from the title of time axis. Sometimes acquisition timezone and time with seconds/millisec precision can be given in  a Log text at the end of the file, which – when present - is loaded and added in the `NDDataset.description`. So if such indications are present, the user can use them – if needed - to improve the date/time of his data. 
- Matlab : when the .mat file is a datasetoject, the data is clearly defined in file specifications (.date field, six-element date/time stamp [year, month, day, hour, minute, second], indicating when the DataSet object was created.  There is no specification for generic .mat files

Text formats
------ 
- JCAMP: well documented in the the file format specification, with two types of date formats: `##DATE :  YY/MM/DD`  in old format versions (established in 1988, and did not take into account the Y2K bug 😊), and `#LONGDATE : YYYY/MM/DD [HH:MM:SS[.SSSS] [±UUUU]]` for JCAMP-DX = 5.01 or higher), which aslo indude optional fractinal seconds and timezone difference with GMT. 

- Csv : no specification for generic .csv files

- Quadera `.asc` : to my knowledge the ris no open file specificatuion, but todate, the date-time format is :  "%m/%d/%Y %H:%M:%S.%f"

- Labspec : it seems that there are two formats for the #Date field : DD.MM.YYYY HH:SS or DD/MM/YYYY HH.SS . In the files present in spectrochempy_data there are 3  file types :
   * SMC1-Initial RT.txt : single spectra with `##Date=	DD.MM.YYYY HH:SS` and `#Acquired=	DD.MM.YYYY HH:SS` fields – this is the case of 532nm-191216-Si 200mu.txt
   * multiple spectra  with `##Date=	DD/MM/YYYY HH:SS` and `#Imported	= DD.MM.YYYY HH:SS` fields – this is the case of Activation.txt 
   * multiple spectra  with `##Date=	DD.MM.YYYY HH:SS`  and `#Acquired	= DD.MM.YYYY HH:SS` fields – this is the case of serie 190214-1.txt
 
Guillaume could have more info about labspec .txt files :-) ?

**Reply by fernandezc** [reply]
Thanks. It will be helpful. 
I have already fixed the problems reading labspec file with different date format.

For information I am trying to convert all datetime handling to np.datetime64 and timedelta64. This for 2 reasons:

- Useful to simplify writing of NetCDF files (and have a full compatibly of the save format with Xarray.

- Second, to be able to have an axis with coordinates of datetime64 dtype. This will advantageously replace the use of labels (when it is just to have the date indication) and will allow plotting with date labels easily when the coordinates are with this dtype.

All Datetimes has to be in GMT because numpy Datetimes are timezone naive. But it is always possible to do simple calculation on the data to change this for a display I think. WIP! will see if it is convenient or not for the v0.4 release.

**Reply by atravert** [reply]
Thank you, I think it would be indeed an improvement ! (time axis with timestamp are not very appealing :-),

For the time zone issue, we would follow these guidelines ?:
- when a date-time is known but timezone is not (grams, labspec, ...) : assume that it is expressed in user's current timezone. This is probably the majority of cases, except of course when the data are acquired and processed in different timezones. In such a case, a possibility would be to add a `timezone` argument in the reader, which default would be the system timezone ?).  
- when the GMT/UTC date-time time is known, but timezone is not (omnic, ...): assume that the timezone is the system time zone -- which is consistent with the previous one.. and also add a  `timezone` argument in the reader with the same default
- when date-time and timezone are known: all fine

**Reply by fernandezc** [reply]
> a possibility would be to add a timezone argument in the reader, which default would be the system timezone 

Good idea! I will add a new metadata in the dataset for this information.
